File: crates/contracts/src/policy.rs
//! Simulation policy contract (simulation-policy.schema.json).

use serde::{Deserialize, Serialize};

use crate::Decimal;

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub struct BalancePolicy {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub insufficient_balance: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub struct SlippagePolicy {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub bps: Option<Decimal>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub struct FeePolicy {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub bps: Option<Decimal>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub struct FillsPolicy {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub partial_fills: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub price_selection: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub slippage: Option<SlippagePolicy>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub fees: Option<FeePolicy>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub struct GasFallback {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub amount: Option<Decimal>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub struct GasPolicy {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub model: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub fallback: Option<GasFallback>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub struct SignalPolicy {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub trigger: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub repeat: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cooldown: Option<String>,
    /// Maximum number of fires when `repeat = "max_count"`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max_count: Option<u32>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub struct OrderingPolicy {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub same_tick: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub struct DataPolicy {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub missing_required: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub missing_optional: Option<String>,
    /// Maximum age (duration like "24h") a venue-scoped mark may be carried
    /// forward when valuing a position (#119(b)). Absent = unbounded.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max_mark_staleness: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub struct PerpPolicy {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub liquidation_check: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub funding: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reduce_only_validation: Option<String>,
    /// Maintenance margin as a fraction of mark notional (e.g. "0.0125").
    /// A position is liquidated once its equity (margin + unrealized PnL)
    /// falls to this fraction of notional. "0" = liquidate only at full
    /// bankruptcy (#120).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub maintenance_margin_ratio: Option<Decimal>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, Default)]
pub struct YieldPolicy {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub accrual: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct SimulationPolicy {
    #[serde(default = "default_policy_schema_version")]
    pub schema_version: String,
    pub profile: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub balance: Option<BalancePolicy>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub fills: Option<FillsPolicy>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub gas: Option<GasPolicy>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub signals: Option<SignalPolicy>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub ordering: Option<OrderingPolicy>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub data: Option<DataPolicy>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub perps: Option<PerpPolicy>,
    #[serde(rename = "yield", default, skip_serializing_if = "Option::is_none")]
    pub yield_: Option<YieldPolicy>,
}

fn default_policy_schema_version() -> String {
    "catalyst.backtest.policy.v1".to_string()
}

File: packages/contracts/src/catalyst_contracts/policy.py
"""Simulation policy contract models (simulation-policy.schema.json)."""

from __future__ import annotations

from typing import Literal

from pydantic import Field

from ._base import Decimal, StrictModel

PolicyProfile = Literal["strict_v1", "conservative_v1", "research_v1"]


class BalancePolicy(StrictModel):
    insufficient_balance: Literal[
        "reject", "partial_fill", "clamp_to_available", "allow_negative"
    ] = "reject"


class SlippagePolicy(StrictModel):
    model: Literal["fixed_bps", "volume_based", "amm_price_impact", "none"] = "fixed_bps"
    bps: Decimal = "10"


class FeePolicy(StrictModel):
    model: Literal["fixed_bps", "venue_fee_table", "none"] = "fixed_bps"
    bps: Decimal = "5"


class FillsPolicy(StrictModel):
    partial_fills: Literal["none", "allow_if_configured", "always_allow"] = "none"
    price_selection: Literal["close", "open", "mid", "next_open", "worse_side_ohlc"] = "close"
    slippage: SlippagePolicy = Field(default_factory=SlippagePolicy)
    fees: FeePolicy = Field(default_factory=FeePolicy)


class GasFallback(StrictModel):
    model: Literal["none", "fixed_usd", "fixed_native"] = "fixed_usd"
    amount: Decimal = "0.25"


class GasPolicy(StrictModel):
    model: Literal["none", "fixed_usd", "fixed_native", "historical_fee_history"] = (
        "historical_fee_history"
    )
    fallback: GasFallback = Field(default_factory=GasFallback)


class SignalPolicy(StrictModel):
    trigger: Literal["level", "crossing", "crossing_with_cooldown", "once_per_backtest"] = (
        "crossing"
    )
    repeat: Literal["never", "on_each_signal_fire", "with_cooldown", "max_count"] = (
        "on_each_signal_fire"
    )
    cooldown: str | None = None


class OrderingPolicy(StrictModel):
    same_tick: Literal[
        "graph_order",
        "topological_order",
        "signals_first_then_actions",
        "conservative_adverse_order",
    ] = "topological_order"


class DataPolicy(StrictModel):
    missing_required: Literal["fail", "skip_tick", "forward_fill"] = "fail"
    missing_optional: Literal["warn", "fail", "forward_fill", "fallback_provider"] = "warn"


class PerpPolicy(StrictModel):
    liquidation_check: Literal["every_tick", "never"] = "every_tick"
    funding: Literal["historical", "none"] = "historical"
    reduce_only_validation: Literal["strict", "lenient"] = "strict"


class YieldPolicy(StrictModel):
    accrual: Literal["simple_apr", "compound_apy", "protocol_index"] = "simple_apr"


class SimulationPolicy(StrictModel):
    schema_version: str = "catalyst.backtest.policy.v1"
    profile: PolicyProfile
    balance: BalancePolicy = Field(default_factory=BalancePolicy)
    fills: FillsPolicy = Field(default_factory=FillsPolicy)
    gas: GasPolicy = Field(default_factory=GasPolicy)
    signals: SignalPolicy = Field(default_factory=SignalPolicy)
    ordering: OrderingPolicy = Field(default_factory=OrderingPolicy)
    data: DataPolicy = Field(default_factory=DataPolicy)
    perps: PerpPolicy = Field(default_factory=PerpPolicy)
    yield_: YieldPolicy = Field(default_factory=YieldPolicy, alias="yield")

    model_config = {"populate_by_name": True, "extra": "forbid"}


__all__ = ["SimulationPolicy", "PolicyProfile"]
